### PR TITLE
Improved notebook; Spot Instance Support; Code for single/multi-CPU/GPU

### PR DESCRIPTION
**Pull request by miroenev**
- Improved notebook text and figures
- Combined all key choices in preamble 
[ dataset size, code type, algorithm choice, cv folds, max_jobs, max_parallel_jobs, search strategy, HPO ranges ]
- Added spot instance support
- Added instance recommendation function based on configuration choices
- Added code branches for 
   * `single-CPU` ( pandas + sklearn )
   * `multi-CPU` ( dask + pandas + sklearn )
   * `single-GPU` ( cudf + cuml )
   * `multi-GPU` ( dask + cudf + cuml )

### CODE REVIEW

File: aws/README.md [hunk 1]
```diff
@@ -27,4 +27,4 @@
    > Note: If you see Pending to the right of the notebook instance in the Status column, your notebook is still being created. The status will change to InService when the notebook is ready for use.
 
 3. Run Notebook
```
Comment (JohnZed) [inline aws/README.md:29]: Maybe note somewhere that it won't work in Studio currently?

File: aws/code/rapids_cloud_ml.py [hunk 2]
```diff
@@ -15,289 +15,450 @@
...
+            parser.add_argument( '--n_estimators', type = int,   default = default_n_estimators )            
+            parser.add_argument( '--max_features', type = float, default = .25 )
+
+            args, unknown_args = parser.parse_known_args( input_args )
+
+            model_params = {            
+                'max_depth' : args.max_depth,
+                'n_estimators' : args.n_estimators,        
+                'max_features': args.max_features,
+                'seed' : 0,
```
Comment (JohnZed) [inline aws/code/rapids_cloud_ml.py:154]: Would be nice to allow users to pass in a fixed seed.

File: aws/code/rapids_cloud_ml.py [hunk 3]
```diff
@@ -15,289 +15,450 @@
...
-                    columns = None  ):
+            raise Exception(f"!error: unknown model type {self.model_type}")
+
+        pprint.pprint( model_params, indent = 5 ); print( '\n' )
+        return model_params
+    
+    # -------------------------------------------------------------------------------------------------------------
+    # ETL
+    # -------------------------------------------------------------------------------------------------------------
+    airline_label_column = 'ArrDel15'
```
Comment (JohnZed) [inline aws/code/rapids_cloud_ml.py:165]: Should we separate out the airline-specific code a bit more from the independent code?

File: aws/notebooks/helper_functions.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
+import random, uuid
+
+def new_job_name_from_config ( dataset_directory, code_choice, 
+                               algorithm_choice, cv_folds,
+                               instance_type, trim_limit = 32 ):
+    job_name = None
```
Comment (JohnZed) [inline aws/notebooks/helper_functions.py:6]: It would be nice to have docstrings since these are user accessible

File: aws/notebooks/hpo.ipynb [hunk 5]
```diff
@@ -0,0 +1,1047 @@
...
+    "max_parallel_jobs = 10"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "max_jobs = 100"
```
Comment (zronaghi) [inline aws/notebooks/hpo.ipynb:283]: Probably better if default is less than 100, can add a comment for larger jobs

File: aws/README.md [hunk 6]
```diff
@@ -27,4 +27,4 @@
...
-   - Once inside JupyterLab you should be able to navigate to the notebook in the **aws/notebooks** directory named **sagemaker_rapids_hpo.ipynb**
+   - Once inside JupyterLab you should be able to navigate to the notebook in the **aws/notebooks** directory named **hpo.ipynb**
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/README.md:30]: I'll go back to the old name, I was thinking it may be overly verbose but it does add redundancy/clarity.

File: aws/code/rapids_cloud_ml.py (hunk 3 above)
Comment (miroenev) [reply aws/code/rapids_cloud_ml.py:165]: Good idea, moved dataset feature and label column definitions to train.py. Hopefully this encourages more BYOData uses.

File: aws/notebooks/helper_functions.py (hunk 4 above)
Comment (miroenev) [reply aws/notebooks/helper_functions.py:6]: Done

File: aws/README.md (hunk 1 above)
Comment (miroenev) [reply aws/README.md:29]: Done

File: aws/code/rapids_cloud_ml.py (hunk 2 above)
Comment (miroenev) [reply aws/code/rapids_cloud_ml.py:154]: Done

File: aws/code/rapids_cloud_ml.py [hunk 7]
```diff
@@ -15,289 +15,450 @@
...
+        self.best_score = -1 
+
+    def enumerate_dataset( self, compute_type, dataset_path, dataset_structure ):
+        # in the case of pandas read-parquet we need to remove the wildcards
+        if 'single-CPU' in compute_type:
+            target_files = dataset_structure['train_data'] + str(dataset_path).split('*')[0]
+        else:
+            target_files = dataset_structure['train_data'] + str(dataset_path)
+                
+        n_datafiles = len( glob.glob( target_files ) )
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/rapids_cloud_ml.py:97]: This was more of a sanity check, to make sure that the bucket you've pointed to is holding the data you are expecting. I can add this additional context as a comment.

Are you OK with keeping this check in the code?

File: aws/notebooks/hpo.ipynb (hunk 5 above)
Comment (miroenev) [reply aws/notebooks/hpo.ipynb:283]: Done

**Comment by miroenev** [comment]
- Finished large XGBoost run and added acceleration + cost savings factors
- Added cluster teardown between folds [ limit memory creep for high CV counts]
- Removed dependence on airline dataset from ETL code [ moved to train.py ]
- Changed RF model save/persistence to use joblib.dump
- Added docstrings
- Added comment in README.md regarding lack of a SageMaker Studio version
- Added seed for XGBoost and RandomForest parsable parameter

**Comment by zronaghi** [comment]
Thanks, looks good to me!
